File: modules/anthra-alarm/android/src/main/java/com/anthra/timer/alarm/AlarmStore.kt
package com.anthra.timer.alarm

import android.content.Context
import org.json.JSONArray
import org.json.JSONObject
import java.util.UUID

data class AlarmConfig(
  val id: Int,
  val label: String,
  val hour: Int,
  val minute: Int,
  val days: List<Int>,
  val pushupTarget: Int,
  val soundUri: String,
  val soundName: String,
  val enabled: Boolean,
  val timezone: String = AlarmStore.TIMEZONE,
  val requiresPushups: Boolean = true
) {
  fun toJson(): JSONObject = JSONObject().apply {
    put("id", id)
    put("label", label)
    put("hour", hour)
    put("minute", minute)
    put("days", JSONArray(days))
    put("pushupTarget", pushupTarget)
    put("soundUri", soundUri)
    put("soundName", soundName)
    put("enabled", enabled)
    put("timezone", timezone)
    put("requiresPushups", requiresPushups)
  }

  companion object {
    fun fromJson(json: JSONObject): AlarmConfig {
      val id = json.getInt("id")
      val rawDays = json.optJSONArray("days") ?: JSONArray()
      val days = buildList {
        for (index in 0 until rawDays.length()) {
          val day = rawDays.optInt(index, -1)
          if (day in 0..6 && !contains(day)) add(day)
        }
      }.ifEmpty { listOf(0, 1, 2, 3, 4, 5, 6) }

      return AlarmConfig(
        id = id,
        label = json.optString("label", "Push-up alarm").take(80),
        hour = json.optInt("hour", 7).coerceIn(0, 23),
        minute = json.optInt("minute", 0).coerceIn(0, 59),
        days = days,
        pushupTarget = json.optInt("pushupTarget", 10).coerceIn(1, 100),
        soundUri = json.optString("soundUri", ""),
        soundName = json.optString("soundName", "System alarm").take(120),
        enabled = json.optBoolean("enabled", true),
        timezone = json.optString("timezone", AlarmStore.TIMEZONE).ifBlank { AlarmStore.TIMEZONE }.take(80),
        // Workout alarms used reserved negative IDs before this field existed.
        // Preserve them as regular alarms across app upgrades and reboots.
        requiresPushups = if (json.has("requiresPushups")) json.optBoolean("requiresPushups", id > 0) else id > 0
      )
    }
  }
}

object AlarmStore {
  const val TIMEZONE = "Asia/Kolkata"
  const val EXTRA_ALARM_ID = "anthra_alarm_id"
  const val EXTRA_FIRED_AT = "anthra_alarm_fired_at"
  const val EXTRA_TEST_MODE = "anthra_alarm_test_mode"
  const val EXTRA_TEST_TARGET = "anthra_alarm_test_target"
  const val EXTRA_DISMISS = "anthra_alarm_dismiss"

  private const val PREFS = "anthra.pushup.alarms.v1"
  private const val KEY_ALARMS = "alarms"
  private const val KEY_COMPLETIONS = "completion_events"

  private fun prefs(context: Context) = context.getSharedPreferences(PREFS, Context.MODE_PRIVATE)

  @Synchronized
  fun save(context: Context, config: AlarmConfig) {
    val root = readAlarmRoot(context)
    root.put(config.id.toString(), config.toJson())
    prefs(context).edit().putString(KEY_ALARMS, root.toString()).apply()
  }

  @Synchronized
  fun remove(context: Context, alarmId: Int) {
    val root = readAlarmRoot(context)
    root.remove(alarmId.toString())
    prefs(context).edit().putString(KEY_ALARMS, root.toString()).apply()
  }

  @Synchronized
  fun clearAlarms(context: Context) {
    prefs(context).edit().putString(KEY_ALARMS, "{}").apply()
  }

  fun get(context: Context, alarmId: Int): AlarmConfig? {
    val root = readAlarmRoot(context)
    val json = root.optJSONObject(alarmId.toString()) ?: return null
    return runCatching { AlarmConfig.fromJson(json) }.getOrNull()
  }

  fun all(context: Context): List<AlarmConfig> {
    val root = readAlarmRoot(context)
    return buildList {
      val keys = root.keys()
      while (keys.hasNext()) {
        val json = root.optJSONObject(keys.next()) ?: continue
        runCatching { AlarmConfig.fromJson(json) }.getOrNull()?.let(::add)
      }
    }
  }

  @Synchronized
  fun addCompletion(
    context: Context,
    config: AlarmConfig,
    firedAt: Long,
    completedReps: Int,
    status: String
  ) {
    // Reserved negative IDs belong to workout reminders, not Alarm Buddy.
    // They should not appear in the Alarm Buddy completion history.
    if (config.id < 0) return
    val entries = readCompletions(context)
    entries.put(JSONObject().apply {
      put("eventId", UUID.randomUUID().toString())
      put("alarmId", config.id.takeIf { it > 0 } ?: JSONObject.NULL)
      put("label", config.label)
      put("firedAt", firedAt)
      put("completedAt", System.currentTimeMillis())
      put("targetReps", config.pushupTarget)
      put("completedReps", completedReps.coerceAtLeast(0))
      put("status", if (status == "emergency_stopped") "emergency_stopped" else "completed")
    })

    val trimmed = JSONArray()
    val start = (entries.length() - 200).coerceAtLeast(0)
    for (index in start until entries.length()) trimmed.put(entries.optJSONObject(index))
    prefs(context).edit().putString(KEY_COMPLETIONS, trimmed.toString()).apply()
  }

  @Synchronized
  fun consumeCompletions(context: Context): JSONArray {
    val entries = readCompletions(context)
    prefs(context).edit().remove(KEY_COMPLETIONS).commit()
    return entries
  }

  private fun readAlarmRoot(context: Context): JSONObject = runCatching {
    JSONObject(prefs(context).getString(KEY_ALARMS, "{}") ?: "{}")
  }.getOrElse { JSONObject() }

  private fun readCompletions(context: Context): JSONArray = runCatching {
    JSONArray(prefs(context).getString(KEY_COMPLETIONS, "[]") ?: "[]")
  }.getOrElse { JSONArray() }
}

File: modules/anthra-alarm/ios/AlarmStore.swift
import Foundation

struct AlarmConfig: Codable {
  var id: Int
  var label: String
  var hour: Int
  var minute: Int
  var days: [Int]
  var pushupTarget: Int
  var soundUri: String
  var soundName: String
  var enabled: Bool
  var timezone: String
  var requiresPushups: Bool

  static let defaultTimezone = "Asia/Kolkata"

  init(
    id: Int,
    label: String,
    hour: Int,
    minute: Int,
    days: [Int],
    pushupTarget: Int,
    soundUri: String,
    soundName: String,
    enabled: Bool,
    timezone: String = defaultTimezone,
    requiresPushups: Bool = true
  ) {
    self.id = id
    self.label = label
    self.hour = hour
    self.minute = minute
    self.days = days.isEmpty ? [0, 1, 2, 3, 4, 5, 6] : days
    self.pushupTarget = min(100, max(1, pushupTarget))
    self.soundUri = soundUri
    self.soundName = soundName
    self.enabled = enabled
    self.timezone = timezone.isEmpty ? Self.defaultTimezone : timezone
    self.requiresPushups = requiresPushups
  }

  init?(dictionary: [String: Any]) {
    guard let id = dictionary["id"] as? Int else { return nil }
    let rawDays = dictionary["days"] as? [Int] ?? []
    let days = rawDays.filter { (0...6).contains($0) }
    self.id = id
    self.label = ((dictionary["label"] as? String)?.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty == false
      ? dictionary["label"] as? String
      : "Push-up alarm") ?? "Push-up alarm"
    self.hour = min(23, max(0, dictionary["hour"] as? Int ?? 7))
    self.minute = min(59, max(0, dictionary["minute"] as? Int ?? 0))
    self.days = days.isEmpty ? [0, 1, 2, 3, 4, 5, 6] : Array(Set(days)).sorted()
    self.pushupTarget = min(100, max(1, dictionary["pushupTarget"] as? Int ?? 10))
    self.soundUri = dictionary["soundUri"] as? String ?? ""
    self.soundName = dictionary["soundName"] as? String ?? "System alarm"
    self.enabled = dictionary["enabled"] as? Bool ?? true
    self.timezone = (dictionary["timezone"] as? String)?.isEmpty == false
      ? dictionary["timezone"] as? String ?? Self.defaultTimezone
      : Self.defaultTimezone
    if dictionary.keys.contains("requiresPushups") {
      self.requiresPushups = dictionary["requiresPushups"] as? Bool ?? (id > 0)
    } else {
      self.requiresPushups = id > 0
    }
  }

  func toDictionary() -> [String: Any] {
    [
      "id": id,
      "label": label,
      "hour": hour,
      "minute": minute,
      "days": days,
      "pushupTarget": pushupTarget,
      "soundUri": soundUri,
      "soundName": soundName,
      "enabled": enabled,
      "timezone": timezone,
      "requiresPushups": requiresPushups
    ]
  }
}

struct AlarmCompletionRecord: Codable {
  let eventId: String
  let alarmId: Int?
  let label: String
  let firedAt: Int64
  let completedAt: Int64
  let targetReps: Int
  let completedReps: Int
  let status: String
}

enum AlarmStore {
  static let firedAtKey = "anthra_alarm_fired_at"
  static let alarmIdKey = "anthra_alarm_id"
  static let testModeKey = "anthra_alarm_test_mode"
  static let testTargetKey = "anthra_alarm_test_target"
  static let dismissKey = "anthra_alarm_dismiss"

  private static let prefsSuite = "anthra.pushup.alarms.v1"
  private static let alarmsKey = "alarms"
  private static let completionsKey = "completion_events"

  private static var defaults: UserDefaults {
    UserDefaults(suiteName: prefsSuite) ?? .standard
  }

  static func save(_ config: AlarmConfig) {
    var root = readAlarmRoot()
    root[String(config.id)] = config.toDictionary()
    if let data = try? JSONSerialization.data(withJSONObject: root),
       let json = String(data: data, encoding: .utf8) {
      defaults.set(json, forKey: alarmsKey)
    }
  }

  static func remove(_ alarmId: Int) {
    var root = readAlarmRoot()
    root.removeValue(forKey: String(alarmId))
    if let data = try? JSONSerialization.data(withJSONObject: root),
       let json = String(data: data, encoding: .utf8) {
      defaults.set(json, forKey: alarmsKey)
    }
  }

  static func clearAlarms() {
    defaults.set("{}", forKey: alarmsKey)
  }

  static func get(_ alarmId: Int) -> AlarmConfig? {
    guard let dict = readAlarmRoot()[String(alarmId)] as? [String: Any] else { return nil }
    return AlarmConfig(dictionary: dict)
  }

  static func all() -> [AlarmConfig] {
    readAlarmRoot().compactMap { _, value in
      guard let dict = value as? [String: Any] else { return nil }
      return AlarmConfig(dictionary: dict)
    }
  }

  static func addCompletion(
    config: AlarmConfig,
    firedAt: Int64,
    completedReps: Int,
    status: String
  ) {
    if config.id < 0 { return }
    var entries = readCompletions()
    entries.append([
      "eventId": UUID().uuidString,
      "alarmId": config.id > 0 ? config.id : NSNull(),
      "label": config.label,
      "firedAt": firedAt,
      "completedAt": Int64(Date().timeIntervalSince1970 * 1000),
      "targetReps": config.pushupTarget,
      "completedReps": max(0, completedReps),
      "status": status == "emergency_stopped" ? "emergency_stopped" : "completed"
    ])
    let trimmed = Array(entries.suffix(200))
    if let data = try? JSONSerialization.data(withJSONObject: trimmed),
       let json = String(data: data, encoding: .utf8) {
      defaults.set(json, forKey: completionsKey)
    }
  }

  static func consumeCompletions() -> [[String: Any]] {
    let entries = readCompletions()
    defaults.removeObject(forKey: completionsKey)
    return entries
  }

  private static func readAlarmRoot() -> [String: Any] {
    guard let json = defaults.string(forKey: alarmsKey),
          let data = json.data(using: .utf8),
          let root = try? JSONSerialization.jsonObject(with: data) as? [String: Any] else {
      return [:]
    }
    return root
  }

  private static func readCompletions() -> [[String: Any]] {
    guard let json = defaults.string(forKey: completionsKey),
          let data = json.data(using: .utf8),
          let root = try? JSONSerialization.jsonObject(with: data) as? [[String: Any]] else {
      return []
    }
    return root
  }
}
